### DISCUSSION TITLE
How to resolve the module imports of a dynamically loaded python extension?

### DISCUSSION THREAD

**Discussion by szabolcsdombi** [pyodide/pyodide#3671] [Q&A]
Hi.

I compiled my rendering library with pyodide build and implemented the necessary glue code for GLES3.0 to WebGL2 in python using Pyodide's `js` module.

So far my examples work well ([poc](https://szabolcsdombi.github.io/zengl-webgl-demo/)).

I would like to skip this and call directly into javascript.
For that I think the easiest would be to just use the module imports like:

```c++
extern void zengl_test_hello();
```

and provide the proper implementation runtime.

Right now Pyodide crashes when I call my function.
How can I add a JS implementation to my module imports?

**Comment by ryanking13** [comment]
Could you give us a minimal reproducible exmple that crashes? I think it should work.

**Comment by szabolcsdombi** [comment]
Full code for the minimal example: https://github.com/szabolcsdombi/pyodide-custom-module-imports

The built extension has an external function called that is not automatically resolved.

```c++
#include <Python.h>

extern void function_implemented_in_javascript(int argument);

PyObject * mymodule_meth_hello(PyObject * self, PyObject * args, PyObject * kwargs) {
    const char * keywords[] = {"arg", NULL};
    PyObject * arg = NULL;
    if (!PyArg_ParseTupleAndKeywords(args, kwargs, "|O", (char **)keywords, &arg)) {
        return NULL;
    }
    printf("calling function_implemented_in_javascript\n");
    function_implemented_in_javascript(1);
    function_implemented_in_javascript(2);
    function_implemented_in_javascript(3);
    printf("success!!!\n");
    Py_RETURN_NONE;
}
```

In the code above `printf` is automatically resolved. How can I manually resolve `function_implemented_in_javascript`?


```js
const pyodide = await loadPyodide();
await pyodide.loadPackage('mymodule-0.1.0-cp310-cp310-emscripten_3_1_27_wasm32.whl');
pyodide.runPython(`import mymodule`);

// pyodide._module.____.____ = (argument) => console.log('hello from JS', argument);
// or do I need the pyodide.FS to access the mymodule.wasm and fill the imports?

document.getElementById('mybutton').addEventListener('click', () => {
  pyodide.runPython(`mymodule.hello('world')`);
});
```

**Reply by hoodmane** [reply]
You need to `#include <emscripten.h>` and then use [the `EM_JS` macro](https://emscripten.org/docs/api_reference/emscripten.h.html#c.EM_JS).

**Comment by szabolcsdombi** [comment]
it now says `#0 14.82 em++: error: EM_JS is not supported in side modules`.

**Reply by hoodmane** [reply]
Okay I will look into this.

**Reply by hoodmane** [reply]
This makes it work:
```patch
diff --git a/index.html b/index.html
index 5af9587..9e257a0 100644
--- a/index.html
+++ b/index.html
@@ -13,6 +13,7 @@
   (async () => {
     const pyodide = await loadPyodide();
     await pyodide.loadPackage('mymodule-0.1.0-cp310-cp310-emscripten_3_1_27_wasm32.whl');
+    pyodide._module.mergeLibSymbols({function_implemented_in_javascript(x) { console.log("Hi from JS", x)}});
     pyodide.runPython(`import mymodule`);
     document.getElementById('mybutton').addEventListener('click', () => {
       pyodide.runPython(`mymodule.hello('world')`);
diff --git a/mymodule.cpp b/mymodule.cpp
index 7eac0db..e0db362 100644
--- a/mymodule.cpp
+++ b/mymodule.cpp
@@ -1,6 +1,6 @@
 #include <Python.h>
 
-extern void function_implemented_in_javascript(int argument);
+extern "C" void function_implemented_in_javascript(int argument);
 
 PyObject * mymodule_meth_hello(PyObject * self, PyObject * args, PyObject * kwargs) {
     const char * keywords[] = {"arg", NULL};
```

**Reply by szabolcsdombi** [reply]
This works 🚀 

Thank you very much.

**Reply by hoodmane** [reply]
It would be nice to fix `EM_JS` support for side modules though. I have an idea of how to implement this, I'll make a branch with an exploration of this.

**Reply by szabolcsdombi** [reply]
emscripten has this [limitation](https://emscripten.org/docs/compiling/Dynamic-Linking.html#limitations)

**Reply by hoodmane** [reply]
Yes. Basically we need two steps:
1. If a wheel has a `.lib.js` file we should load it in our loader logic
2. If a package uses `EM_JS` we should emit a `.lib.js` file

The latter requires patching emscripten so that instead of bailing when it sees `em_js` symbols in a side module it generates the appropriate loader logic. I'll start with step 1.

**Reply by szabolcsdombi** [reply]
does this also mean if I link my side module with -sFULL_ES3 and have the necessary glue code in the .lib.js it would work out of the box with pyodide?

**Reply by hoodmane** [reply]
Eventually something like that would probably be desirable, but it is outside the scope of what I'm currently experimenting with.

**Comment by szabolcsdombi** [comment]
Thank you very much for the help.

I have separated my glue code to a js file and load it with the mergeLibSymbols

This way my render time got down to match native rendertime.
Previously this example did not hit 60fps, but now a frame time is around 3-4ms way below 16.6ms (60fps).

- [live demo](https://szabolcsdombi.github.io/zengl-webgl-demo/food.html)
- [source](https://github.com/szabolcsdombi/zengl/blob/c3a158826ce3491ead5a90ad98d6257f13a1bb19/examples/webgl/public/food.html)

I would be more than happy to test the `.lib.js` auto loading variant once available.
It sounds perfect for my use-case.
